app/services/experimentation/data_export.rb

```diff
@@ -0,0 +1,195 @@
+            communications: process_communications(notifications),
+            patient_gender: patient.gender,
+            patient_age: patient.age,
+            patient_risk_level: patient.risk_priority,
+            diagnosed_hypertensive: true, # remove?
+            patient_has_phone: true, # remove?
+            assigned_facility_name: assigned_facility&.name,
+            assigned_facility_type: assigned_facility&.facility_type,
+            assigned_facility_state: assigned_facility&.state,
+            assigned_facility_district: assigned_facility&.district,
+            assigned_facility_block: assigned_facility&.block,
+            patient_registration_date: patient.device_created_at.to_date,
+            patient_id: tgm.id
+          }
+        end
+      end
+    end
+
+    def process_communications(notifications)
+      communications = notifications.each_with_object([]) do |notification, communications_data|
+        ordered_communications = notification.communications.sort_by{|c| c.created_at }
+        ordered_communications.each do |c|
+          communications_data << [notification.message, c.communication_type, c.detailable&.delivered_on, c.detailable&.result]
+        end
+      end
+
+      @max_communications = communications.count if communications.count > @max_communications
+      communications
+    end
+
+    def pad_communications
+      aggregate.each do |patient_data|
+        communications_deficit = @max_communications - patient_data[:communications].count
+        communications_deficit.times do
+          patient_data[:communications] << Array.new(4, nil)
+        end
+      end
+    end
+
+    def process_appointments(patient, notifications)
+      appts = if notifications.any?
+        notifications.map(&:subject).uniq.sort_by{|appt| appt.scheduled_date }
+      else
+        date_range = experiment.start_date.beginning_of_day..experiment.end_date.end_of_day
+        patient.appointments.where(scheduled_date: date_range).where("device_created_at < ?", notification_start_date).order(:scheduled_date)
+      end
+      @max_appointments = appts.count if appts.count > max_appointments
+
+      encounters_during_experiment = encounter_dates(patient, notification_start_date, cutoff_date)
+
+      appts.each_with_index.map do |appt, index|
+        # pulling out encounters sequentially because there's no formal relationship to appointments
+        followup_date = encounters_during_experiment[index]
+        days_til_followup = nil
+        bp_at_followup = nil
+        facility = nil
+
+        if followup_date
+          days_til_followup = (appt.scheduled_date - followup_date).to_i
+          encounter = encounter_by_date(patient, followup_date)
+          bp_at_followup = encounter.class == BloodPressure
+          facility = encounter.facility
+        end
+        [appt.device_created_at.to_date, appt.scheduled_date, followup_date, days_til_followup, bp_at_followup,
```

Comment:
Using arrays instead of hashes hurts readability slightly but also reduces some complexity. I think the values being added are clear enough, but it's still not ideal.